### ISSUE TITLE
Confidence scores for matched documents

### ISSUE THREAD

**Issue by davemoore-** [zentity-io/zentity#27]
**Background**

Many users have requested a way to indicate the confidence of a match for documents returned in the output of a resolution job. Often the request is for a score, where a higher score indicates greater confidence in the match. Some users envisioned a score for each document. Others envisioned a score for specific fields such as the value of a name.

Currently, and with no change in mind for the future, zentity submits boolean queries to find matching documents in Elasticsearch. Therefore, by the standards of Elasticsearch, every matching document has a constant score of 1.

zentity also offers an [`"_explanation"`](https://zentity.io/docs/entity-resolution/output-specification/#hits.hits._explanation) field for each document, which describes the resolvers, matchers, and values that caused the document to match.

I believe it would be possible to let users assign scores for various concepts in zentity, to combine those scores to produce an overall confidence score for a matching document, and to implement this in a way that intuitively fits the design of zentity and does not incur a significant performance penalty.

@cmwaters89 deserves recognition for demonstrating the feasibility of this concept. Thank you for your contribution!

**Concept**

The following zentity components could be extended to support a user-defined score that contributes to an overall confidence score for any matching document:

- Attribute
- Resolver
- Matcher
- Index
- Index Field
- Hop

Examples:

- **Attributes** - Users might observe that some attributes, like `ssn` or `email`, are likely to identify a single entity more accurately than other attributes like `name` or `dob`.
- **Resolvers** - Likewise, some combinations of attributes (i.e. "resolvers") are likely to identify a single entity more accurately than others.
- **Matchers** - Exact matchers could be considered more confident than fuzzy, phonetic, or ranged matchers.
- **Indices** - Some indices could be known to have data that's more or less reliable (clean, accurate, trustworthy) than others.
- **Index Fields** - Likewise, some index fields in particular could be known to have data that's more or less reliable than others.
- **Hops** - Documents found in later hops could be considered less confident than documents found in earlier hops.

The concept of the feature is to let users to define a base score for any of the components listed above in whatever way they see fit for their use case. These base scores could be defined in the entity model and overridden at query time.

Upon query execution, zentity would then, for each document, combine the component base scores to produce an overall match confidence score for the document. This may involve calculating the product or average of the base scores. The way in which zentity combines the scores could also be configured by the user.

**Default behavior**

The default score for any document should be 1, null, or simply not present in the document at all, both to remain agnostic and to be consistent with past versions of the plugin.

The default base score for any component listed above should not influence the overall confidence score of the document. It should only contribute to that score when the user defines the base score. Potentially one solution for this is to let the default base score be null (i.e. undefined) and to skip any such scores when calculating the overall score.

**Deconflicting base scores**

It's possible for documents to match for multiple reasons. A document could match multiple resolvers and matchers. A case could be made for doing either of the following:

1. Use the base score for the best resolver or matcher.
2. Use the combined base scores for every resolver or matcher.

This behavior could be configured by the user, too.

**Score thresholds**

It should be easy enough to allow the user to defind a threshold that a document score must pass in order to be kept in the results. This has been another common and related feature request.

**Optional and opt-in**

The implementation of this feature would depend on information from the [`"_explanation"`](https://zentity.io/docs/entity-resolution/output-specification/#hits.hits._explanation) field. This information is only gathered when the client sets `_explanation=true` in the URL, because the queries become slightly more complex and incur a slight performance penalty. Likewise, I believe this scoring feature should be made optional and "opt-in" for the users who care about it.

**Generic**

One of the tenants of zentity is to be generic and not domain-specific. zentity should remain agnostic to the actual scoring process, and instead provide a framework in which the user can define a scoring process that fits their use case.

**Scoring based on past scores**

If a document matches with a relatively low score, should its values penalize the scores of documents that match it in subsequent queries? I have yet to think this through in detail.

**Field level scores**

This feature should work at the document level. I'm not confident (no pun intended) that this could be done at the field level without adding a lot of complexity to the plugin. By field level, I mean a score that indicates the confidence that the value of a specific field matches an input value. zentity is able to explain the reason for a match by using named filters in Elasticsearch. Named filters can inform zentity of the matchers that led to a hit, but it doesn't provide details on why they led to a hit.

Users who desire field level scores have a couple alternatives:

1. Have the client application take the `"input_value"` and `"target_value"` from the `"_explanation"` field to derive a confidence score for the value outside of zentity. Perhaps the client application derives a score by comparing the length and edit distance of the two values, for example.
2. Upon the release of this feature, treat the score from the matcher as a proxy for the score for the value, since the matcher already describes why two values match.

**Comment by davemoore-** [comment]
I'm back from a break and ready to begin this feature.

Below is my plan for an initial release of this feature. I won't immediately implement every concept from my prior post, but I'll try to future-proof the implementation so that the feature can mature without introducing breaking changes. The implementation will be simple for users. This post documents the complexity that will be solved behind the scenes.

**Summary**

The Resolution API will support an option to return a confidence score (`"_score"`) for each document in the result. The scores will allow documents in the result to be ranked by a measurement of confidence. Users would define base scores and modifiers for attributes, matchers, and index fields in the entity model. zentity would use those to calculate a confidence score for each matching document by referencing the attributes, matchers, and index fields in the `"_explanation"` field of each document.

**Document confidence score**

A **document confidence score** represents the confidence that a document matches the entity of the resolution job. The score is a floating point number in the range of `0.0` - `1.0`, although in practice the scores should be in the range of `0.5` - `1.0` as I'll soon explain. Any score in between the special values of `0.0`, `0.5`, and `1.0` is arbitrary. The purpose of the score is to rank the matching documents by a measurement of confidence.

- A score of `1.0` represents 100% confidence that the document is a match.
- A score of `0.0` represents 100% confidence that the document is not a match.
- A score of `0.5` represents 100% uncertainty that the document is a match.
- A score of `null` indicates that the document lacks a confidence score.

A document confidence score is calculated from three sets of data values:

- The base scores of the matching attributes of the document.
- The quality score of the matcher for each matching attribute of the document.
- The quality score of the index field for each matching attribute of the document.

The user defines the base scores of the attributes and the quality scores of the matchers and index fields in the entity model. The resolution job calculates document confidence scores by applying those base scores and quality scores to the values of the `"_explanation"` fields of document in a resolution job.

**Attribute confidence score**

An **attribute confidence score** represents the confidence that a matching attribute uniquely identifies the entity after accounting for the quality of the matcher and the index field. The score is a floating point number in the range of `0.0` - `1.0` and has a similar meaning as the document match score.

- A score of `1.0` represents 100% confidence that the attribute identifies the entity.
- A score of `0.0` represents 100% confidence that the attribute does not identify the entity.
- A score of `0.5` represents 100% uncertainty that the attribute identifies the entity.
- A score of `null` indicates that the attribute lacks a base score.

Effectively, if a document matches with one or more attributes:

- The document score always will be `1.0` if any attribute has a match score of `1.0`.
- The document score always will be `0.0` if any attribute has a match score of `0.0`.
- The document score will be unaffected by any attribute with a match score of `0.5`.
- The document score will be unaffected by any attribute with a match score of `null`.

**Attribute base score**

An **attribute base score** represents the confidence that an attribute would uniquely identify the entity if it were to match, assuming the quality of its matcher and index field are perfect. An attribute base score is hypothetical, whereas an attribute confidence score is an applied measurement of confidence using actual data as inputs.

Generally it makes sense for every attribute to have a base score between `0.5` - `1.0`. A base score that's less than `0.5` would indicate that the matching attribute represents some level of a false match, which is contrary to the general usage of zentity where a matching attribute represents some level of a true match.

Cardinality would be a good statistic by which to define an attribute base score. For example:

- "High" cardinality attributes such as `"ssn"` or `"phone"` or `"ip"` should have a base score closer to `1.0` because they're more likely to uniquely identify the entity.
- "Low" cardinality attributes such as `"sex"` should have a base score closer to `0.5` because they don't contribute significantly to the identification of the entity.
- "Normal" cardinality attributes such as `"first_name"` or `"street"` or `"dob"` should have a base score somewhere in between the two extremes of `0.5` and `1.0`.

Care should be taken when using an attribute base score of `1.0` or `0.0`, because it would allow a single attribute confidence score to determine the document confidence score. Whenever there is an attribute confidence score of `1.0` or `0.0`, it takes precedence over any other attribute confidence score in the document. For example, you might have an `"id"` field that you absolutely trust to identify an entity. If you allow the attribute confidence score of the `"id"` field to be `1.0`, then anytime the `"id"` field matches in a document, no other attribute confidence score would matter because you've already stated that the `"id"` field always indicates a match with perfect confidence. A best practice would be to use a high number such as `0.99` to allow for some small level of variability and more nuanced rankings of documents.

**Matcher quality score**

A **matcher quality score** represents the quality or trustworthiness of a matcher. It modifies the attribute base score and contributes to the final attribute confidence score.

- A quality score of `1.0` represents 100% confidence that the matcher can be trusted.
- A quality score of `0.0` represents 100% confidence that the matcher cannot be trusted.
- A quality score of `null` indicates that the matcher lacks a quality score.

Effectively this means:

- A quality score of `1.0` will not affect the attribute base score.
- A quality score of less than `1.0` will penalize the attribute base score.
- A quality score of `0.0` will set the attribute base score to `0.5`.
- A quality score of `null` will not affect the attribute base score.

The purpose of the matcher quality score is to reflect any dubious matcher quality in the final document confidence score. For example, an exact matcher may have a quality score of `1.0`, while a fuzzy matcher may have a quality score of `0.95` to express slightly less confidence in the quality of the match.

**Index field quality score**

An **index field quality score** represents the quality or trustworthiness of the data in an index field. It modifies the attribute base score and contributes to the final attribute confidence score.

- A quality score of `1.0` represents 100% confidence that the index field data can be trusted.
- A quality score of `0.0` represents 100% confidence that the index field data cannot be trusted.
- A quality score of `null` indicates that the index field lacks a quality score.

Effectively this means:

- A quality score of `1.0` will not affect the attribute base score.
- A quality score of less than `1.0` will penalize the attribute base score.
- A quality score of `0.0` will set the attribute base score to `0.5`.
- A quality score of `null` will not affect the attribute base score.

The purpose of the index field quality score is to reflect any dubious data quality in the final document confidence score. For example, an index field with perfectly clean and governed data may have a quality score of `1.0`, while an index field with known data quality issues may have a quality score of `0.95` to express slightly less confidence in the quality of the match.

**Calculating an attribute confidence score**

A matcher quality score and an index field quality score both modify an attribute base score in the same way. The attribute base score is multiplied by the quality score, and the product then is normalized on a scale from `0.5` to the value of the attribute base score. Effectively this means that a perfect quality score of `1.0` will not affect the attribute base score, which is desirable because the attribute base score already assumes a hypothetically perfect condition of quality. Likewise a quality score of `0.0` will reduce the attribute base score to `0.5`, which is desirable because it represents total uncertainty given the total lack of matcher quality or data quality, and it will not affect the final document confidence score.

Here is the formula by which a quality score modifies an attribute base score:

`((attribute_base_score - 0.5) / (attribute_base_score - 0.0) * ((attribute_base_score * quality_score) - attribute_base_score)) + attribute_base_score`

Examples:

- When `attribute_base_score = 0.75` and `quality_score = 1.00`, then `attribute_confidence_score = 0.7500`
- When `attribute_base_score = 0.75` and `quality_score = 0.90`, then `attribute_confidence_score = 0.7250`
- When `attribute_base_score = 0.75` and `quality_score = 0.75`, then `attribute_confidence_score = 0.6875`
- When `attribute_base_score = 0.75` and `quality_score = 0.50`, then `attribute_confidence_score = 0.6250`
- When `attribute_base_score = 0.75` and `quality_score = 0.25`, then `attribute_confidence_score = 0.5625`
- When `attribute_base_score = 0.75` and `quality_score = 0.10`, then `attribute_confidence_score = 0.5250`
- When `attribute_base_score = 0.75` and `quality_score = 0.00`, then `attribute_confidence_score = 0.5000`

The formula above is calculated once for each quality score. The first calculation uses the matcher quality score. The second calculation uses the index field quality score while substituting the `attribute_base_score` with the output of the prior calculation. The order of these calculations does not affect the final attribute confidence score.

Any quality scores of `null` are omitted from the calculation.

**Deconflicting attribute confidence scores for multiple matches of the same attribute**

It's possible for a document to have more than one matching value for the same attribute type. If there are multiple matches for the same attribute, then the maximum attribute confidence score for that attribute will be used to calculate the final document confidence score.

**Calculating the document confidence score**

The final document confidence score is calculated from one or more attribute confidence scores. The attribute confidence scores are combined using [conflation of probability distributions](https://arxiv.org/pdf/0808.1808v4.pdf). This algorithm treats each attribute confidence score as independent evidence. When there is more evidence, the confidence score becomes higher than the individual scores. This is desirable because a greater number of matching attributes implies greater evidence (thus greater confidence) of a true match.

Here is the formula by which the document confidence score is calculated from attribute confidence scores:

`document_confidence_score = product_of_scores / (product_of_scores + product_of_inversed_scores)`

Here is a concrete example with two attribute confidence scores `0.65` and `0.85`:

`document_confidence_score = (.65 * .85) / (.65 * .85 + ((1-.65) * (1-.85)))`
`document_confidence_score = 0.91322314049`

Examples:

- When `attribute_confidence_scores = [ 0.75, 1.00 ]`, then `document_confidence_score = 1.0`
- When `attribute_confidence_scores = [ 0.75, 0.95 ]`, then `document_confidence_score = 0.98275862069`
- When `attribute_confidence_scores = [ 0.75, 0.85 ]`, then `document_confidence_score = 0.94444444444`
- When `attribute_confidence_scores = [ 0.55, 0.65, 0.75 ]`, then `document_confidence_score = 0.87195121951`
- When `attribute_confidence_scores = [ 0.55, 0.65, 0.75, 0.50 ]`, then `document_confidence_score = 0.87195121951`
- When `attribute_confidence_scores = [ 0.75, 0.00 ]`, then `document_confidence_score = 0.0`
- When `attribute_confidence_scores = [ 1.00, 0.00 ]`, then `document_confidence_score = zero_division_error`

Any attribute confidence scores of `null` are omitted from the calculation.

**Pseudocode for the entire scoring process**

```python
# Calculate the attribute confidence scores for each attribute.
index_name = doc["_index"]
attribute_confidence_scores = {}
for match in doc["_explanation"]["matches"]:
    attribute_name = match["attribute"]
    matcher_name = match["input_matcher"]
    index_field_name = match["target_field"]
    if attribute_name not in attribute_confidence_scores:
        attribute_confidence_scores[attribute_name] = []

    # Modify the attribute base score with the quality scores.
    abs = model["attributes"][attribute_name]["score"] # Attribute base score
    mqs = model["matchers"][matcher_name]["quality"] # Matcher quality score
    iqs = model["indices"][index_name]["fields"][index_field_name]["quality"] # Index field quality score
    tmp = abs # Temporary running attribute confidence score
    tmp = ((tmp - 0.5) / (tmp - 0.0) * ((tmp * mqs) - tmp)) + tmp
    tmp = ((tmp - 0.5) / (tmp - 0.0) * ((tmp * iqs) - tmp)) + tmp
    attribute_confidence_score = abs # Final attribute confidence score
    attribute_confidence_scores[attribute_name].append(attribute_confidence_score)

# Deconflict multiple attribute confidence scores for the same attribute by
# selecting the highest score.
for attribute_name in attribute_confidence_scores:
    attribute_confidence_scores[attribute_name] = max(attribute_confidence_scores[attribute_name])

# Combine the attribute confidence scores into a final document confidence score
# using conflation of probability distributions.
scores = []
scores_inverse = []
for attribute_name in attribute_confidence_scores:
  score = attribute_confidence_scores[attribute_name]
  scores.append(score)
  scores_inverse.append(1.0 - score)
product_scores = product(scores)
product_scores_inverse = product(scores_inverse)
document_confidence_score = product_scores / (product_scores + product_scores_inverse)

return document_confidence_score
```

**Additions to the Entity Model**

Three optional fields will be added to the entity model:

- `"attribute"."ATTRIBUTE_NAME"."score"` - The base score for an attribute.
- `"matchers"."MATCHER_NAME"."quality"` - A base score modifier that represents the quality of the matcher.
- `"indices"."INDEX_NAME"."fields"."INDEX_FIELD_NAME"."quality"` A base score modifier that represents the quality of the index field's data.

Example:

```javascript
{
  "attributes": {
    ATTRIBUTE_NAME: {
      "score": FLOAT,
      "type": ...
    },
    ...
  },
  "matchers": {
    MATCHER_NAME: {
      "clause": {
        ...
      },
      "quality": FLOAT
    },
    ...
  },
  "indices": {
    INDEX_NAME: {
      "fields": {
        INDEX_FIELD_NAME: {
          "attribute": ...,
          "matcher": ...,
          "quality": FLOAT
        }
      }
    }
  }
}
```

**Additions to the Resolution API**

One optional URL parameter will be added to the Resolution API:

- `_score` - Boolean - Defaults to `false` to optimize performance.

**Comment by davemoore-** [comment]
I've released [zentity-1.6.0](https://zentity.io/releases/#zentity-1.6.0) which implements the scoring feature of this issue.

Here's the official nomenclature compared to my earlier proposal:

- **Composite identity confidence score** replaces "document confidence score"
- **Attribute identity confidence score** replaces "attribute confidence score"
- **Attribute identity confidence base score** replaces "attribute base score"
- **Matcher quality score** remains the same
- **Index field quality score** remains the same

The algorithm I've implemented is nearly identical to the pseudocode in my earlier proposal. It includes proper handling of division by zero.

What remains to be done:

- [x] Document the new fields in the [Resolution API](https://zentity.io/docs/rest-apis/resolution-api), [Entity Model Specification](https://zentity.io/docs/entity-models/specification), and [Entity Resolution Output Specification](https://zentity.io/docs/entity-resolution/output-specification).
- [ ] Update the [zentity sandbox](https://zentity.io/sandbox) to use zentity-1.6.0 with example scores in the entity models.
- [ ] Publish a tutorial on scoring in the documentation.

**Comment by seanlindo** [comment]
This is great stuff, Dave. Thank you!
